File: app/src/main/java/za/co/jpsoft/winkerkreader/CallLogAdapter.java
package za.co.jpsoft.winkerkreader;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

public class CallLogAdapter extends RecyclerView.Adapter<CallLogAdapter.CallLogViewHolder> {

    private List<CallLog> callLogs;

    public CallLogAdapter(List<CallLog> callLogs) {
        this.callLogs = callLogs;
    }

    public static class CallLogViewHolder extends RecyclerView.ViewHolder {
        public TextView callIcon;
        public TextView callerName;
        public TextView dateTime;
        public TextView callNumber;

        public CallLogViewHolder(View itemView) {
            super(itemView);
            callIcon = itemView.findViewById(R.id.callIcon);
            callerName = itemView.findViewById(R.id.callerName);
            dateTime = itemView.findViewById(R.id.dateTime);
            callNumber = itemView.findViewById(R.id.callNumber);
        }
    }

    @NonNull
    @Override
    public CallLogViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_call_log, parent, false);
        return new CallLogViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull CallLogViewHolder holder, int position) {
        CallLog callLog = callLogs.get(position);
        
        // Set appropriate icon based on call type and source
        String icon;
        if ("Phone Call".equals(callLog.getSource())) {
            switch (callLog.getCallType()) {
                case "INCOMING":
                    icon = "📞";
                    break;
                case "OUTGOING":
                    icon = "📤";
                    break;
                case "MISSED":
                    icon = "📵";
                    break;
                default:
                    icon = "📞";
                    break;
            }
        } else if ("WhatsApp".equals(callLog.getSource())) {
            icon = "💬";
        } else {
            icon = "📞";
        }
        
        holder.callIcon.setText(icon);
        holder.callerName.setText(callLog.getCallerInfo());
        holder.dateTime.setText(callLog.getFormattedDateTime());
        
        // Show call type, source, and duration
        StringBuilder callInfo = new StringBuilder();
        callInfo.append(callLog.getCallType()).append(" • ").append(callLog.getSource());
        
        if (callLog.getDuration() > 0) {
            long minutes = callLog.getDuration() / 60;
            long seconds = callLog.getDuration() % 60;
            callInfo.append(" • ").append(minutes).append("m ").append(seconds).append("s");
        }
        
        holder.callNumber.setText(callInfo.toString());
    }

    @Override
    public int getItemCount() {
        return callLogs.size();
    }

    public void updateLogs(List<CallLog> newLogs) {
        callLogs.clear();
        callLogs.addAll(newLogs);
        notifyDataSetChanged();
    }
    
    public void addLog(CallLog callLog) {
        callLogs.add(0, callLog); // Add to beginning for latest first
        notifyItemInserted(0);
    }
}

File: app/src/main/kotlin/za/co/jpsoft/winkerkreader/ui/adapters/CallLogAdapter.kt
package za.co.jpsoft.winkerkreader.ui.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import za.co.jpsoft.winkerkreader.databinding.ItemCallLogBinding
import za.co.jpsoft.winkerkreader.data.models.CallLog

class CallLogAdapter(initialLogs: List<CallLog>) : RecyclerView.Adapter<CallLogAdapter.CallLogViewHolder>() {

    private var callLogs: List<CallLog> = initialLogs.toList()

    class CallLogViewHolder(val binding: ItemCallLogBinding) : RecyclerView.ViewHolder(binding.root)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CallLogViewHolder {
        val binding = ItemCallLogBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return CallLogViewHolder(binding)
    }

    override fun onBindViewHolder(holder: CallLogViewHolder, position: Int) {
        val callLog = callLogs[position]
        val icon = when {
            callLog.source == "Phone Call" -> {
                when (callLog.callType) {
                    "INCOMING" -> "📞"
                    "OUTGOING" -> "📤"
                    "MISSED" -> "📵"
                    else -> "📞"
                }
            }
            callLog.source == "WhatsApp" -> "💬"
            else -> "📞"
        }
        with(holder.binding) {
            callIcon.text = icon
            callerName.text = callLog.callerInfo
            dateTime.text = callLog.formattedDateTime
            val callInfo = StringBuilder().apply {
                append(callLog.callType).append(" • ").append(callLog.source)
                if (callLog.duration > 0) {
                    val minutes = callLog.duration / 60
                    val seconds = callLog.duration % 60
                    append(" • ").append(minutes).append("m ").append(seconds).append("s")
                }
            }
            callNumber.text = callInfo.toString()
        }
    }

    override fun getItemCount() = callLogs.size

    fun updateLogs(newLogs: List<CallLog>) {
        val diffResult = DiffUtil.calculateDiff(CallLogDiffCallback(callLogs, newLogs))
        callLogs = newLogs.toList()
        diffResult.dispatchUpdatesTo(this)
    }

    private class CallLogDiffCallback(
        private val oldList: List<CallLog>,
        private val newList: List<CallLog>
    ) : DiffUtil.Callback() {
        override fun getOldListSize() = oldList.size
        override fun getNewListSize() = newList.size
        override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            return oldList[oldItemPosition].id == newList[newItemPosition].id
        }
        override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            return oldList[oldItemPosition] == newList[newItemPosition]
        }
    }
}